File: sdks/node/src/webhooks.ts
/**
 * Webhook signature verification (M22.4).
 *
 * ## Why this is the most important function in the package
 *
 * Everything else here is convenience: an integrator who skipped this SDK entirely would still
 * end up with working payments, just more slowly. This one is different — a receiver that does
 * not verify signatures will accept a forged `payment.captured` from anyone who learns the URL,
 * and a receiver that verifies the *body* but ignores the *timestamp* will accept a genuine
 * delivery replayed forever. §4.5 calls the timestamp the detail homegrown implementations most
 * often get wrong, and this function exists so that nobody has to get it right twice.
 *
 * ## The specification, which is M18's and not this SDK's
 *
 * ```
 * PaymentFlow-Signature: t=1785758400,v1=5f2c…9ab
 *
 *   signed_payload = "{t}" + "." + "{raw request body}"
 *   v1             = lowercase hex of HMAC-SHA256(secret, signed_payload)
 *   secret         = the endpoint's whsec_… value as UTF-8 bytes, prefix included
 * ```
 *
 * Several `v1` values may appear in one header, comma-separated: that is how the dual-secret
 * rotation window is expressed on the wire, so a receiver that has already switched and one
 * that has not both verify. Any match is a match.
 *
 * This implementation is checked against
 * `notification-service/src/test/resources/signature-vectors/webhook-signature-vectors.json`
 * — the same five vectors the platform's own signer, and the reference `verify.js` and
 * `verify.py`, are checked against. That file's header comment anticipates this SDK consuming
 * it, so a divergence fails a test here rather than being discovered by a merchant.
 */

import { createHmac, timingSafeEqual } from 'node:crypto';
import {
  WebhookPayloadError,
  WebhookSignatureError,
  WebhookTimestampError,
} from './errors.js';

/** The header every delivery carries. */
export const SIGNATURE_HEADER = 'PaymentFlow-Signature';

/**
 * The default tolerance window, in seconds.
 *
 * Five minutes, which is what the merchant-facing guide recommends. Wide enough to survive
 * ordinary clock drift between two servers, narrow enough that a captured delivery is not
 * replayable for the rest of the afternoon.
 */
export const DEFAULT_TOLERANCE_SECONDS = 300;

/**
 * The event envelope a verified delivery contains.
 *
 * Hand-written rather than reused from the generated models, because the two are genuinely
 * different shapes: `/v1/events` returns `EventResponse`, which has no `apiVersion`, while a
 * delivery does carry one. Returning the generated type would under-declare the object this
 * function actually hands back.
 */
export interface WebhookEvent {
  /** `evt_` followed by 32 hex characters. Stable across retries and replays — **dedupe on it**. */
  readonly id: string;
  /** Always `event`. */
  readonly object: string;
  /**
   * What happened: `payment.created`, `payment.authorized`, `payment.captured`,
   * `payment.failed`, `payment.voided`, `payment.refunded`, `payment.partially_refunded`.
   *
   * Typed as `string` and not a union on purpose. New event types are additive and ship
   * without a new API revision, so a handler must ignore what it does not recognise rather
   * than fail on it — a closed union would make the compiler enforce the opposite.
   */
  readonly type: string;
  /** The dated API revision the payload is shaped by. */
  readonly apiVersion?: string;
  /** When the event occurred, as RFC 3339. Not when it was delivered. */
  readonly created?: string;
  /** `test` or `live`. */
  readonly mode?: string;
  /**
   * The resource the event happened to, as it was at the time.
   *
   * `data.object` rather than `data` directly: the wrapper is the seam that lets a later
   * revision add siblings — `previousAttributes` being the obvious one — without changing the
   * type of `data`, which would be breaking. Its shape depends on `type` and is deliberately
   * open here, because a webhook handler branches on `type` before it reads the object.
   */
  readonly data: { readonly object?: Record<string, unknown> };
}

/** Options for {@link constructEvent}, for callers who prefer names to positions. */
export interface ConstructEventOptions {
  /** How far a delivery's timestamp may be from now, in seconds. Defaults to 300. */
  readonly toleranceSeconds?: number;
  /** The current time in epoch seconds. For tests; defaults to the system clock. */
  readonly nowEpochSeconds?: number;
}

/**
 * Verifies a delivery and returns its event.
 *
 * @param payload the **raw** request body, exactly as received. See the warning below.
 * @param signatureHeader the value of the `PaymentFlow-Signature` header.
 * @param secret the endpoint's `whsec_…` signing secret.
 * @param tolerance seconds of allowed clock skew, or an options object. Defaults to 300.
 *
 * @throws {WebhookSignatureError} the header is malformed, or nothing in it matched.
 * @throws {WebhookTimestampError} the signature is valid and its timestamp is out of window.
 * @throws {WebhookPayloadError} it verified and is not an event envelope.
 *
 * ## The raw body, and why it has to be raw
 *
 * The signature covers the bytes that were sent. `JSON.parse` followed by `JSON.stringify`
 * does not round-trip them — key order, whitespace and number formatting are all free to
 * change — so a caller who passes a re-serialized object will get a signature failure on a
 * delivery that was perfectly valid. In Express that means `express.raw({ type: 'application/json' })`
 * on this route, *before* `express.json()` sees it. This is the single most common way a
 * correct integration fails, which is why it is the first thing said here.
 *
 * A `Buffer` or `Uint8Array` is accepted directly and is the safer thing to pass, because it
 * cannot have been decoded and re-encoded on the way.
 */
export function constructEvent(
  payload: string | Uint8Array,
  signatureHeader: string,
  secret: string,
  tolerance: number | ConstructEventOptions = DEFAULT_TOLERANCE_SECONDS,
): WebhookEvent {
  const options: ConstructEventOptions = typeof tolerance === 'number' ? { toleranceSeconds: tolerance } : tolerance;
  const toleranceSeconds = options.toleranceSeconds ?? DEFAULT_TOLERANCE_SECONDS;
  const now = options.nowEpochSeconds ?? Math.floor(Date.now() / 1000);

  if (typeof secret !== 'string' || secret.length === 0) {
    throw new WebhookSignatureError('No signing secret. Pass the endpoint\'s `whsec_…` value.');
  }
  if (typeof signatureHeader !== 'string' || signatureHeader.length === 0) {
    throw new WebhookSignatureError(`No ${SIGNATURE_HEADER} header on the request.`);
  }
  if (!Number.isFinite(toleranceSeconds) || toleranceSeconds < 0) {
    throw new WebhookSignatureError('`tolerance` must be a non-negative number of seconds.');
  }

  const { timestamp, candidates } = parseHeader(signatureHeader);

  // Order matters, and this order is deliberate: the signature is checked *before* the
  // timestamp. Checking the window first would let anyone with the URL and a stopwatch learn
  // whether a body was correctly signed by observing which error came back — and it would
  // report a garbage header as "too old", which sends an integrator to look at their clock.
  const body = typeof payload === 'string' ? Buffer.from(payload, 'utf8') : Buffer.from(payload);
  const expected = sign(secret, timestamp, body);
  if (!candidates.some((candidate) => constantTimeEquals(candidate, expected))) {
    throw new WebhookSignatureError(
      'The signature does not match. Either the secret is wrong, or the payload is not the raw ' +
        'request body — re-serializing the JSON changes the bytes the signature covers.',
    );
  }

  const skew = Math.abs(now - timestamp);
  if (skew > toleranceSeconds) {
    // Absolute skew, not just "too old". A timestamp far in the future is equally a sign the
    // header was not produced for this delivery at this moment.
    throw new WebhookTimestampError(
      `The delivery's timestamp is ${skew}s away from now, outside the ${toleranceSeconds}s ` +
        'tolerance. This is a replayed delivery, or one of the two clocks is wrong.',
      timestamp,
      skew,
    );
  }

  return parseEvent(body.toString('utf8'));
}

/**
 * Computes the `v1` value for a body: lowercase hex HMAC-SHA256 over `"{timestamp}.{body}"`.
 *
 * Exported so a caller can build a signed request in their own tests without reimplementing
 * the specification from the guide — which is the moment they would get it subtly wrong, and
 * then write a test that passes against their own mistake.
 */
export function signPayload(secret: string, timestampEpochSeconds: number, payload: string | Uint8Array): string {
  const body = typeof payload === 'string' ? Buffer.from(payload, 'utf8') : Buffer.from(payload);
  return sign(secret, timestampEpochSeconds, body);
}

/** Builds a full header value, for the same reason {@link signPayload} is exported. */
export function signatureHeaderFor(
  secret: string,
  timestampEpochSeconds: number,
  payload: string | Uint8Array,
): string {
  return `t=${timestampEpochSeconds},v1=${signPayload(secret, timestampEpochSeconds, payload)}`;
}

// ── Internals ───────────────────────────────────────────────────────────────────────────

function sign(secret: string, timestamp: number, body: Buffer): string {
  const hmac = createHmac('sha256', Buffer.from(secret, 'utf8'));
  hmac.update(Buffer.from(`${timestamp}.`, 'utf8'));
  hmac.update(body);
  return hmac.digest('hex');
}

interface ParsedHeader {
  readonly timestamp: number;
  readonly candidates: readonly string[];
}

function parseHeader(header: string): ParsedHeader {
  let timestamp: number | undefined;
  const candidates: string[] = [];

  for (const element of header.split(',')) {
    const separator = element.indexOf('=');
    if (separator < 0) continue;
    const key = element.slice(0, separator).trim();
    const value = element.slice(separator + 1).trim();
    if (key === 't') {
      // Not `Number(value)`: that accepts '', '0x10' and '1e9', so a header with an empty
      // timestamp would parse as epoch zero and be reported as fifty years of clock skew.
      if (!/^\d+$/.test(value)) {
        throw new WebhookSignatureError(`The ${SIGNATURE_HEADER} header's timestamp is not an integer.`);
      }
      timestamp = Number(value);
    } else if (key === 'v1') {
      if (value.length > 0) candidates.push(value);
    }
    // Unknown fields are skipped rather than rejected. A future `v2` alongside `v1` is how
    // this scheme would gain a second algorithm, and a verifier that refused the whole header
    // on sight of an unfamiliar field would break on the day that shipped.
  }

  if (timestamp === undefined) {
    throw new WebhookSignatureError(`The ${SIGNATURE_HEADER} header has no \`t=\` timestamp.`);
  }
  if (candidates.length === 0) {
    throw new WebhookSignatureError(`The ${SIGNATURE_HEADER} header has no \`v1=\` signature.`);
  }
  return { timestamp, candidates };
}

/**
 * Compares two hex signatures without leaking where they first differ.
 *
 * `timingSafeEqual` throws on a length mismatch, so the length is checked first — which does
 * leak the length, and cannot leak anything else: a `v1` is always 64 hex characters, so a
 * candidate of any other length is malformed rather than a near miss.
 */
function constantTimeEquals(candidate: string, expected: string): boolean {
  const a = Buffer.from(candidate, 'utf8');
  const b = Buffer.from(expected, 'utf8');
  return a.length === b.length && timingSafeEqual(a, b);
}

function parseEvent(raw: string): WebhookEvent {
  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch {
    throw new WebhookPayloadError('The delivery verified but its body is not JSON.');
  }
  if (typeof parsed !== 'object' || parsed === null || Array.isArray(parsed)) {
    throw new WebhookPayloadError('The delivery verified but its body is not a JSON object.');
  }

  const event = parsed as Partial<WebhookEvent>;
  // Exactly the three fields this function's return type promises, and no more. Validating
  // the rest would break §9's forward-compatibility promise the first time a field was added.
  if (typeof event.id !== 'string' || typeof event.type !== 'string' || typeof event.data !== 'object' || event.data === null) {
    throw new WebhookPayloadError(
      'The delivery verified but is not an event envelope — `id`, `type` and `data` are required.',
    );
  }
  return event as WebhookEvent;
}

File: sdks/go/webhooks.go
package paymentflow

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"regexp"
	"strconv"
	"strings"
	"time"
)

// SignatureHeader is the header every webhook delivery carries.
const SignatureHeader = "PaymentFlow-Signature"

// DefaultTolerance is the default clock-skew window: five minutes. Wide enough to survive
// ordinary drift between two servers, narrow enough that a captured delivery is not replayable
// for the rest of the afternoon.
const DefaultTolerance = 5 * time.Minute

// WebhookEvent is the envelope a verified delivery contains. It is not model.Event: /v1/events
// returns no apiVersion, a delivery does. ID (evt_ + 32 hex) is stable across retries and
// replays — dedupe on it. Data["object"] is the resource the event happened to; branch on Type
// before reading it.
type WebhookEvent struct {
	ID         string         `json:"id"`
	Object     string         `json:"object"`
	Type       string         `json:"type"`
	APIVersion string         `json:"apiVersion"`
	Created    string         `json:"created"`
	Mode       string         `json:"mode"`
	Data       map[string]any `json:"data"`
}

// DataObject returns Data["object"] as a map, or nil.
func (e WebhookEvent) DataObject() map[string]any {
	if e.Data == nil {
		return nil
	}
	obj, _ := e.Data["object"].(map[string]any)
	return obj
}

// WebhookSignatureError: the signature header was malformed, or nothing in it matched. Treat it
// as hostile — a body that fails verification did not come from PaymentFlow, or did not arrive
// intact, and either way must not be acted on.
type WebhookSignatureError struct{ Message string }

func (e *WebhookSignatureError) Error() string { return "paymentflow: " + e.Message }

// WebhookTimestampError: the signature was valid and its timestamp is outside the tolerance
// window. A different problem from a signature failure, with a different fix — usually a replay,
// but also what a clock skewed by minutes looks like.
type WebhookTimestampError struct {
	Message     string
	Timestamp   int64
	SkewSeconds int64
}

func (e *WebhookTimestampError) Error() string { return "paymentflow: " + e.Message }

// WebhookPayloadError: it verified and is not an event envelope. Reachable only from a platform
// defect.
type WebhookPayloadError struct{ Message string }

func (e *WebhookPayloadError) Error() string { return "paymentflow: " + e.Message }

var digitsRE = regexp.MustCompile(`^\d+$`)

// ConstructEvent verifies a delivery and returns its event.
//
// payload must be the RAW request body, exactly as received — the signature covers the bytes
// that were sent, and parsing then re-serializing the JSON does not round-trip them. In net/http
// that is the []byte from io.ReadAll(r.Body), read before anything decodes it.
//
// tolerance of 0 uses DefaultTolerance. Pass a negative value and it is rejected.
func ConstructEvent(payload []byte, signatureHeader, secret string, tolerance time.Duration) (*WebhookEvent, error) {
	return constructEventAt(payload, signatureHeader, secret, tolerance, time.Now())
}

func constructEventAt(payload []byte, signatureHeader, secret string, tolerance time.Duration, now time.Time) (*WebhookEvent, error) {
	if tolerance == 0 {
		tolerance = DefaultTolerance
	}
	if secret == "" {
		return nil, &WebhookSignatureError{"No signing secret. Pass the endpoint's whsec_… value."}
	}
	if signatureHeader == "" {
		return nil, &WebhookSignatureError{"No " + SignatureHeader + " header on the request."}
	}
	if tolerance < 0 {
		return nil, &WebhookSignatureError{"tolerance must not be negative."}
	}

	timestamp, candidates, err := parseSignatureHeader(signatureHeader)
	if err != nil {
		return nil, err
	}

	// Signature before timestamp, deliberately: checking the window first would let anyone with
	// the URL and a stopwatch learn whether a body was correctly signed by observing which
	// error came back, and would report a garbage header as "too old".
	expected := sign(secret, timestamp, payload)
	matched := false
	for _, candidate := range candidates {
		if hmac.Equal([]byte(candidate), []byte(expected)) {
			matched = true
			break
		}
	}
	if !matched {
		return nil, &WebhookSignatureError{
			"The signature does not match. Either the secret is wrong, or the payload is not the " +
				"raw request body — re-serializing the JSON changes the bytes the signature covers.",
		}
	}

	skew := now.Unix() - timestamp
	if skew < 0 {
		skew = -skew
	}
	if skew > int64(tolerance.Seconds()) {
		return nil, &WebhookTimestampError{
			Message: fmt.Sprintf("The delivery's timestamp is %ds away from now, outside the %ds tolerance. "+
				"This is a replayed delivery, or a clock is wrong.", skew, int64(tolerance.Seconds())),
			Timestamp:   timestamp,
			SkewSeconds: skew,
		}
	}

	return parseWebhookEvent(payload)
}

// SignPayload computes the v1 value for a body: lowercase hex HMAC-SHA256 over "{timestamp}.{body}".
// Exported so a caller can build a signed request in their own tests without reimplementing the
// specification — the moment they would get it subtly wrong and write a test that passes
// against their own mistake.
func SignPayload(secret string, timestamp int64, payload []byte) string {
	return sign(secret, timestamp, payload)
}

// SignatureHeaderFor builds a full header value, for the same reason SignPayload is exported.
func SignatureHeaderFor(secret string, timestamp int64, payload []byte) string {
	return fmt.Sprintf("t=%d,v1=%s", timestamp, sign(secret, timestamp, payload))
}

func sign(secret string, timestamp int64, body []byte) string {
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(strconv.FormatInt(timestamp, 10) + "."))
	mac.Write(body)
	return hex.EncodeToString(mac.Sum(nil))
}

func parseSignatureHeader(header string) (int64, []string, error) {
	var timestamp int64 = -1
	var candidates []string
	for _, element := range strings.Split(header, ",") {
		sep := strings.IndexByte(element, '=')
		if sep < 0 {
			continue
		}
		key := strings.TrimSpace(element[:sep])
		value := strings.TrimSpace(element[sep+1:])
		switch key {
		case "t":
			if !digitsRE.MatchString(value) {
				return 0, nil, &WebhookSignatureError{"The " + SignatureHeader + " header's timestamp is not an integer."}
			}
			timestamp, _ = strconv.ParseInt(value, 10, 64)
		case "v1":
			if value != "" {
				candidates = append(candidates, value)
			}
		}
		// A future v2 alongside v1 is how this scheme would gain a second algorithm; an
		// unfamiliar field is skipped rather than rejected.
	}
	if timestamp < 0 {
		return 0, nil, &WebhookSignatureError{"The " + SignatureHeader + " header has no t= timestamp."}
	}
	if len(candidates) == 0 {
		return 0, nil, &WebhookSignatureError{"The " + SignatureHeader + " header has no v1= signature."}
	}
	return timestamp, candidates, nil
}

func parseWebhookEvent(body []byte) (*WebhookEvent, error) {
	var generic map[string]any
	if err := json.Unmarshal(body, &generic); err != nil {
		return nil, &WebhookPayloadError{"The delivery verified but its body is not a JSON object."}
	}
	_, idOK := generic["id"].(string)
	_, typeOK := generic["type"].(string)
	_, dataOK := generic["data"].(map[string]any)
	if !idOK || !typeOK || !dataOK {
		return nil, &WebhookPayloadError{
			"The delivery verified but is not an event envelope — id, type and data are required.",
		}
	}
	var event WebhookEvent
	if err := json.Unmarshal(body, &event); err != nil {
		return nil, &WebhookPayloadError{"The delivery verified but its body is not an event envelope."}
	}
	return &event, nil
}
